doc/plp_snprintf.3
.TH "PLP_SNPRINTF" 3 "Dec 12, 1998" V98.12.13
.SH NAME
plp_snprintf, plp_vsnprintf - formatted output conversion
.SH SYNOPSIS
.br
#include <stdarg.h>
.br
void plp_snprintf(char *str, int size, const char *format, ...);
.br
void plp_vsnprintf(char *str, int size, const char *format, va_list ap);
.br
extern int visible_control;
.br
.SH "HISTORY, TERMS AND CONDITIONS"
.IX "plp_snprintf" "" "format and print C language values"
.LP
This version of plp_snprintf was developed originally for printing on a motley
collection of specialized hardware that had NO IO library.
Due to contractual
restrictions,  a clean room implementation of the printf() code had to be
developed.
.LP
The implementation of plp_snprintf tends to be overly paranoid,
as these platforms had NO memory protection,  and very small address spaces.
This made it possible to try to print very long strings, i.e. - all of memory,
very easily.
To guard against this,  all printing was done via a fixed size buffer,
generous enough to hold reasonable strings,  but small enough to protect against
overruns, etc.
.LP
Strangely enough,  this proves to be of immense importance when SPRINTFing to
a buffer on a stack.
The rest,  of course,  is well known,  as buffer overruns
in the stack are a common way to do horrible things to operating systems,
security, etc etc.
.LP
This version of the plp_snprintf documentation is based on the
snprintf() manual page from BSDI Inc,
the POSIX printf() documentation, and the 4.4 BSD Release printf() manual page.
.LP
The plp_snprintf() source code is freely distributible under the following
license terms:
.LP
You may use, copy, distribute, or otherwise incorporate this software and documentation
into any product or other item,  provided that the copyright in the documentation and source code
as well as the source code generated
constant strings in the object, executable or other code remain in place and are present
in executable modules or objects.
.LP
You may modify this
code as appropriate to your usage; however the modified version must be identified by changing
the various source and object code identification strings as is appropriately
noted in the source code.
.LP
.SH DESCRIPTION
.LP
The plp_snprintf() and plp_vsnprintf() procedures write to a string under
the control of a format string that specifies how subsequent arguments
output.
.LP
The plp_snprintf() and plp_vsnprintf() will write at most size-1
characters floowed by a `\0'.
If str is 0 (NULL) or size is 0, then no characters are written.
.LP
The visible_control variable (default value 1) is used when printing character
or string values.  If the character C to be printed is a control character
or and not a space as determined by the ctype.h library iscntrl()
and isspace() functions,  then it will be printed as the form ^X,
where X = (C & 0x1f) | '@' in C language notation.
.LP
The format string is composed of zero or more directives:
ordinary characters (not %), which are copied unchanged to the output stream;
and conversion specifications, each of which results in fetching zero or more
subsequent arguments.
Each conversion specification is introduced by the
character %. The arguments must correspond properly (after type promotion) with the conversion specifier.
After the %, the following appear
in sequence:
.IP o
Zero or more of the following flags:
.LP
.IP -
A zero `0' character specifying zero padding.
For all conversions except n, the converted value is padded on the left with
zeros rather than blanks.
If a precision is given with a numeric
conversion (d, i, o, u, i, x, and X), the `0' flag is ignored.
.LP
.IP -
A negative field width flag `-' indicates the converted value is
to be left adjusted on the field boundary.
.LP
.IP -
A `+' character specifying that a sign always be placed before a
number produced by a signed conversion.
.LP
.IP o
An optional decimal digit string specifying a minimum field width.
If the converted value has fewer characters than the field width, it
will be padded with spaces on the left
(or right, if the left-adjustment flag has been given) to fill out the field width.
Note that e, f, and g formats are restricted to a maximum of 64 characters.
.LP
.IP o
An optional precision, in the form of a period `.' followed by an optional digit string.
If the digit string is omitted, the precision
is taken as zero.
This gives the minimum number of digits to appear
for d, i, o, u, x, and X conversions, the number of digits to appear
after the decimal-point.
For e, f, and g conversions, the meaning depends on the local system sprintf() support.
.LP
.IP o
The optional character l (ell) specifying that a following d, i, o,
u, x, or X conversion applies to a long int or unsigned
long int argument.
.LP
.IP o
The optional character q, specifying that a following d, i, o, u, x,
or X conversion corresponds to a quad or long long.
.LP
.IP o
A character that specifies the type of conversion to be applied.
.LP
A field width or precision, or both, may be indicated by an asterisk `*'
instead of a digit string.
In this case, an int argument supplies the
field width or precision.
A negative or zero field width is ignored.
.LP
The conversion specifiers and their meanings are:
.IP diouxX
The int (or appropriate variant) argument is converted to signed
decimal (d and i), unsigned octal (o), unsigned decimal (u), or
unsigned hexadecimal (x and X) notation.
The letters abcdef are
used for x conversions; the letters ABCDEF are used for X conversions.
The precision, if any, gives the minimum number of digits
that must appear; if the converted value requires fewer digits,
it is padded on the left with blanks unless zero padding has been requested.
.IP efg
The double argument is rounded and converted in the style
of the underlying systems sprintf() support.  Lengths and other critical
values have been check,  and the maximum length of converted value is 64 bytes.
.IP c
The int argument is converted to an unsigned char, and the resulting character is written,
as modified by the visible_control variable value.
.IP s
The ``char *'' argument is expected to be a pointer to an array
of character type (pointer to a string).
Characters from the array are written up to (but not including) a terminating NUL character; if a precision is specified, no more than the number specified are written.
Characters are written to the string as modified by the visible_control variable value.
.IP %
A `%' is written. No argument is converted. The complete conversion specification is `%%'.
.LP
In no case does a non-existent or small field width cause truncation of a
field; if the result of a conversion is wider than the field width, the
field is expanded to contain the conversion result.
.LP
SH BUGS
.lp
The typedef names quad_t and u_quad_t are ugly, ugly, ugly.
Your system may use different quad type names.
.LP
.SH COPYRIGHT
.LP
The source code and documentation is Copyright 1988-1999 by Patrick Powell
<papowell@astart.com>.


doc/CHANGES
-- 2.4.3 release 20181119 "" --

 012) Thomas E. Dickey <dickey@invisible-island.net>
      ADD. update configure macros
      FIX. aclocal.m4, configure[.in], autoconf.h[in], memory.c, screen.c
           string.c, tcurses.c

 011) Francisco Javier Serrador <fserrador@gmail.com>
      ADD. spanish translation
      FIX. es.po

 010) Matt Anton <tin@lv223.org>
      ADD. updated french translation
      FIX. fr.po

 009) Dennis Preiser <dennis@d--p.de>
      BUG. ensure destination of match_string() is NULL terminated
      BUG. possible buffer overflows with long translations
      BUG. fixed length buffer in rfc1522_decode() may overflow
      ADD. don't idna_decode() strings which are not IDNA encoded
      ADD. ability to filter on Path:-header
      FIX. art.c, config.c, feed.c, filter.c, getline.c, help.c, lang.c
           main.c, memory.c, nntplib.c, page.c, rfc2047.c, save.c, screen.c
           extern.h, nntplib.h, proto.h, tin.h, tin.5

 008) Urs Janssen <urs@tin.org>
      ADD. config.guess, config.sub update
      ADD. checks for explicit_bzero(), explicit_memset(), memset_s()
      ADD. check for unicode/unorm2.h to favor unorm2_normalize() over
           unorm_normalize() if found
      ADD. NFKC_case_fold normalization if unicode/unorm2.h is found
      ADD. don't strip binray with "make install"
      ADD. match_string()/match_item() now stop on first '\n'
      ADD. don't show broken active file messages if not running with -D 1
      ADD. accept arbitrary line length in mail.active file
      ADD. allow (multiple) -v for debug mode
      ADD. honor NDEBUG in assert() replacement
      BUG. possible buffer overflow when prompting for a wildcard search
           pattern in get_search_pattern()
      BUG. possible buffer overflows with long translations
      BUG. pointer arithmetic with possible NULL pointer
      BUG. didn't recheck CAPABILITIES after auth
      BUG. transport-padding on mime boundaries wasn't ignored
      FIX. active.c, art.c, auth.c, config.c, cook.c, debug.c, feed.c
           getline.c, group.c, help.c, init.c, keymap.c, lang.c, mail.c, main.c
           misc.c, newsrc.c, nntplib.c, options_menu.c, page.c, post.c, read.c
           rfc1524.c, rfc2045.c, rfc2046.c, rfc2047.c, search.c, screen.c
           select.c, signal.c, strftime.c, string.c, extern.h, proto.h, tin.h
           config.guess, config.sub, configure[.in], autoconf.h[in]
           Makefile[.in], tin.1, tin.5

-- 2.4.2 release 20171224 "Lochhead" --

 007) Thomas E. Dickey <dickey@invisible-island.net>
      ADD. update configure macros
      FIX. aclocal.m4, configure[.in]

 006) Mesutcan Kurt <mesutcank@gmail.com>
      ADD. updated turkish translation
      FIX. tr.po

 005) Dennis Preiser <dennis@d--p.de>
      ADD. cancel_lock_algo config option
      ADD. sha256 and sha512 as possible algorithms for cancel_locks
      ADD. sigfile=!command no longer passes the name of the current
           newsgroup to command, but %G is replaced by groupname if
           specified
      ADD. remove sleep(1) from unshar loop
      BUG. possible buffer overrun in rfc1522_decode()
      BUG. wildcard search doesn't advance past a match in a line for a
           subsequent search
      BUG. last line in the info_pager (help, posting-history) lacked
           truncation
      BUG. quoted pairs in CFWS were not ignored
      ADD. cleanup
      FIX. attrib.c, config.c, inews.c, init.c, lang.c, options_menu.c, page.c
           post.c, regex.c, rfc2045.c, rfc2046.c, rfc2047.c, save.c, sigfile.c
           string.c, wildmat.c, tincfg.tbl, extern.h, tinrc.h, version.h
           libcanlock/*, tin.1, tin.5

 004) Joe Hansen <joedalton2@yahoo.dk>
      ADD. updated danish translation
      FIX. da.po

 003) Urs Janssen <urs@tin.org>
      BUG. possible NULL dereference when IDNA-decoding with libicuuc
      BUG. possible NULL dereference with -D 4
      BUG. possible use of uninitialized variable in debug output
      BUG. posting history screen failed to use additional space
           if window was widened
      BUG. 2nd arg of tin_getline() is int not bool
      BUG. multibyte input to prompts was grabbled if using ncurses but not
           ncursesw
      ADD. config.guess, config.sub update
      ADD. adjust argument names to be the same in declaration and definition
      ADD. install rtin.5 manpage
      ADD. skip group query for -w if group was given on the cmd.-line
      ADD. remove redundant return statement at the end of void functions
      ADD. tweak MAIL_REGEX to not capture a tailing ')'
      ADD. use uidna_nameToUnicode() instead of uidna_IDNToUnicode() if
           linked against ICU >=4.6
      FIX. art.c, attrib.c, auth.c, charset.c, config.c, curses.c, feed.c
           filter.c, group.c, help.c, keymap.c, lang.c, main.c, mimetypes.c
           misc.c, newsrc.c, nntplib.c, pgp.c, post.c, prompt.c, refs.c
           rfc2047.c, save.c, screen.c, select.c, tcurses.c, thread.c, proto.h
           tin.h, Makefile[.in], configure[.in], autoconf.h[in], config.guess
           config.sub, tin.1, tin.5

-- 2.4.1 release 20161224 "Daill" --

 002) Dennis Preiser <dennis@d--p.de>
      BUG. reconnection problem if started with '-n' and multiple
           "LIST COUNTS grp1,grp2,..." were sent
      BUG. overlapping strcpy(3)/wcscpy(3)
      FIX. getline.c, select.c

 001) Urs Janssen <urs@tin.org>
      ADD. config.guess, config.sub update
      ADD. clarify ternary condition
      BUG. logic error when connecting to a server in transit mode forced
           user to auth even if server didn't request it
      BUG. resource leak in post_process_uud() and parse_rfc2045_article()
      BUG. possible nullpointer dereferenced when running out of memory in
           makecfg (during compilation)
      BUG. memleak if using libunistring for unicode normalization
      FIX. active.c, art.c, auth.c, config.c, feed.c, group.c, makecfg.c

src/plp_snprintf.c
/**************************************************************************
 * LPRng IFHP Filter
 * Copyright 1994-1999 Patrick Powell, San Diego, CA <papowell@astart.com>
 **************************************************************************/

/*
 * Overview:
 *
 * This version of snprintf was developed originally for printing
 * on a motley collection of specialized hardware that had NO IO
 * library.  Due to contractual restrictions,  a clean room implementation
 * of the printf() code had to be developed.
 *
 * The method chosen for printf was to be as paranoid as possible,
 * as these platforms had NO memory protection,  and very small
 * address spaces.  This made it possible to try to print
 * very long strings, i.e. - all of memory, very easily.  To guard
 * against this,  all printing was done via a buffer, generous enough
 * to hold strings,  but small enough to protect against overruns,
 * etc.
 *
 * Strangely enough,  this proved to be of immense importance when
 * SPRINTFing to a buffer on a stack...  The rest,  of course,  is
 * well known,  as buffer overruns in the stack are a common way to
 * do horrible things to operating systems, security, etc etc.
 *
 * This version of snprintf is VERY limited by modern standards.
 *
 * COPYRIGHT AND TERMS OF USE:
 *
 * You may use, copy, distribute, or otherwise incorporate this software
 * and documentation into any product or other item,  provided that
 * the copyright in the documentation and source code as well as the
 * source code generated constant strings in the object, executable
 * or other code remain in place and are present in executable modules
 * or objects.
 *
 * You may modify this code as appropriate to your usage; however the
 * modified version must be identified by changing the various source
 * and object code identification strings as is appropriately noted
 * in the source code.
 *
 * The next include line is expected to work in conjunction with the
 * GNU CONFIGURE utility.  You  should define the following macros
 * appropriately:
 *
 * HAVE_STDARG_H - if the <stdargs.h> include file is available
 * HAVE_VARARG_H - if the <varargs.h> include file is available
 *
 * HAVE_STRERROR - if the strerror() routine is available.
 *   If it is not available, then examine the lines containing
 *   the tests below.  You may need to fiddle with HAVE_SYS_NERR
 *   and  HAVE_SYS_NERR_DEF to make compilation work correctly.
 *        HAVE_SYS_NERR
 *        HAVE_SYS_NERR_DEF
 *
 * HAVE_QUAD_T   - if the quad_t type is defined
 * HAVE_LONG_LONG  - if the long long type is defined
 *
 *   If you are using the GNU configure (autoconf) facility, add the
 *   following line to the configure.in file, to force checking for the
 *   quad_t and long long  data types:
 *
 *     AC_CHECK_TYPE(quad_t,NONE)
 *
 * 	dnl test to see if long long is defined
 *
 * 	AC_MSG_CHECKING(checking for long long)
 * 	AC_TRY_COMPILE([
 * 	#include <sys/types.h>
 * 	],[long long x; x = 0],
 * 	ac_cv_long_long=yes, ac_cv_long_long=no)
 * 	AC_MSG_RESULT($ac_cv_long_long)
 * 	if test $ac_cv_long_long = yes; then
 * 	  AC_DEFINE(HAVE_LONG_LONG)
 * 	fi
 *
 *
 *   Add the following lines to the acconfig.h in the correct
 *   position.
 *
 *     / * Define if quad_t is NOT present on the system  * /
 *     #undef quad_t
 *
 *     / * Define if long long is present on the system  * /
 *     #undef HAVE_LONG_LONG
 *
 *   When you run configure, if quad_t is NOT defined, the config.h
 *   file will have in it:
 *     #define quad_t NONE
 *
 *   If it is defined, the config.h file will have
 *   / * #undef quad_t * /
 *
 *   If long long is defined, you will have:
 *     #define HAVE_LONG_LONG
 *
 *    This code is then used in the source code to enable quad quad
 *   and long long support.
 *
 */


#ifndef TIN_H
#	include "tin.h"
#endif /* !TIN_H */

#if !defined(HAVE_SNPRINTF) && !defined(HAVE_VSNPRINTF)

#	if !defined(PLP_SNPRINTF_H) && !defined(TIN_H)
#		include "plp_snprintf.h"
#	endif /* !PLP_SNPRINTF_H && !TIN_H */

/*
 * KEEP THIS STRING - MODIFY AT THE END WITH YOUR REVISIONS
 * i.e. - the LOCAL REVISIONS part is for your use
 */

static const char *_id = "plp_snprintf V1999.02.20 Copyright Patrick Powell 1988-1999 <papowell@astart.com> \
@Id: plp_snprintf.c,v 1.4 1999/02/20 17:44:16 papowell Exp papowell @\
 LOCAL REVISIONS: tin 1.9.5-01";

/* varargs declarations: */

#ifdef HAVE_STDARGS
#	undef HAVE_STDARGS    /* let's hope that works everywhere (mj) */
#endif /* HAVE_STDARGS */
#ifdef VA_LOCAL_DECL
#	undef VA_LOCAL_DECL
#endif /* VA_LOCAL_DECL */
#ifdef VA_START
#	undef VA_START
#endif /* VA_START */
#ifdef VA_SHIFT
#	undef VA_SHIFT
#endif /* VA_SHIFT */
#ifdef VA_END
#	undef VA_END
#endif /* VA_END */

#ifdef HAVE_STDARG_H
#	include <stdarg.h>
#	define HAVE_STDARGS    /* let's hope that works everywhere (mj) */
#	define VA_LOCAL_DECL   va_list ap;
#	define VA_START(f)     va_start(ap, f)
#	define VA_SHIFT(v,t)	;	/* no-op for ANSI */
#	define VA_END          va_end(ap)
#else
#	ifdef HAVE_VARARGS_H
#		include <varargs.h>
#		undef HAVE_STDARGS
#		define VA_LOCAL_DECL   va_list ap;
#		define VA_START(f)     va_start(ap)		/* f is ignored! */
#		define VA_SHIFT(v,t)	v = va_arg(ap,t)
#		define VA_END		va_end(ap)
#	else
XX ** NO VARARGS ** XX
#	endif /* HAVE_VARARGS_H */
#endif /* HAVE_STDARG_H */

#if 0
/* the dreaded QUAD_T strikes again... */
#ifndef HAVE_QUAD_T
#	if defined(quad_t) && qaud_t == NONE
#		define HAVE_QUAD_T 0
#	else
#		define HAVE_QUAD_T 1
#	endif /* quad_t && qaud_t== NONE */
#endif /* HAVE_QUAD_T */
#endif /* 0 */

#if defined(HAVE_QUAD_T) && !defined(HAVE_LONG_LONG)
  ERROR you need long long
#endif
#ifdef HAVE_QUAD_T
   /* suspender and belts on this one */
   const union { quad_t t; long long v; } x;
#endif

union value {
#if defined(HAVE_QUAD_T) || defined(HAVE_LONG_LONG)
	long long value;
#else
	long value;
#endif
	double dvalue;
};

#ifdef cval
#	undef cval
#endif /* cval */
#define cval(s) (*((unsigned const char *)s))


static char *plp_Errormsg(int err);
static void plp_strcat(char *dest, const char *src);
static void dopr(char *buffer, const char *format, va_list args);
static void fmtstr(const char *value, int ljust, int len, int precision);
static void fmtnum(union value *value, int plp_base, int dosign, int ljust, int len, int zpad);
static void fmtdouble(int fmt, double value, int ljust, int len, int zpad, int precision);
static void dostr(const char *);
static void dopr_outch(int c);

static char *output;
static char *end;

int visible_control = 1;


int plp_vsnprintf(char *str, size_t count, const char *fmt, va_list args)
{
	str[0] = 0;
	end = str+count-1;
	dopr( str, fmt, args );
	if( count != 0 )
		end[0] = 0;
	return(strlen(str));
}

/* VARARGS3 */
#ifdef HAVE_STDARGS
int plp_snprintf(char *str,size_t count,const char *fmt,...)
#else
int plp_snprintf(va_alist) va_dcl
#endif
{
#ifndef HAVE_STDARGS
    char *str;
	size_t count;
    char *fmt;
#endif
    VA_LOCAL_DECL

    VA_START (fmt);
    VA_SHIFT (str, char *);
    VA_SHIFT (count, size_t );
    VA_SHIFT (fmt, char *);
    (void) plp_vsnprintf( str, count, fmt, ap);
    VA_END;
	return( strlen( str ) );
}

static void dopr( char *buffer, const char *format, va_list args )
{
	int ch;
	union value value;
	int longflag = 0;
	int quadflag = 0;
	char *strvalue;
	int ljust;
	int len;
	int zpad;
	int precision;
	int set_precision;
	double dval;
	int err = errno;
	int plp_base = 0;
	int signed_val = 0;

	output = buffer;
	while( (ch = *format++) ){
		switch( ch ){
		case '%':
			longflag = quadflag =
			ljust = len = zpad = plp_base = signed_val = 0;
			precision = -1; set_precision = 0;
		nextch:
			ch = *format++;
			switch( ch ){
			case 0:
				dostr( "**end of format**" );
				return;
			case '-': ljust = 1; goto nextch;
			case '.': set_precision = 1; precision = 0; goto nextch;
			case '*':
				if( set_precision ){
					precision = va_arg( args, int );
				} else {
					len = va_arg( args, int );
				}
				goto nextch;
			case '0': /* set zero padding if len not set */
				if(len==0 && set_precision == 0 ) zpad = '0';
				/* FALLTHROUGH */
			case '1': /* FALLTHROUGH */
			case '2': /* FALLTHROUGH */
			case '3': /* FALLTHROUGH */
			case '4': /* FALLTHROUGH */
			case '5': /* FALLTHROUGH */
			case '6': /* FALLTHROUGH */
			case '7': /* FALLTHROUGH */
			case '8': /* FALLTHROUGH */
			case '9':
				if( set_precision ){
					precision = precision*10 + ch - '0';
				} else {
					len = len*10 + ch - '0';
				}
				goto nextch;
			case 'l': ++longflag; goto nextch;
			case 'q': quadflag = 1; goto nextch;
			case 'u': case 'U':
				if( plp_base == 0 ){ plp_base = 10; signed_val = 0; }
				/* FALLTHROUGH */
			case 'o': case 'O':
				if( plp_base == 0 ){ plp_base = 8; signed_val = 0; }
				/* FALLTHROUGH */
			case 'd': case 'D':
				if( plp_base == 0 ){ plp_base = 10; signed_val = 1; }
				/* FALLTHROUGH */
			case 'x':
				if( plp_base == 0 ){ plp_base = 16; signed_val = 0; }
				/* FALLTHROUGH */
			case 'X':
				if( plp_base == 0 ){ plp_base = -16; signed_val = 0; }
				if( quadflag || longflag > 1 ){
#if defined(HAVE_LONG_LONG)
					if( signed_val ){
					value.value = va_arg( args, long long );
					} else {
					value.value = va_arg( args, unsigned long long );
					}
#else
					if( signed_val ){
					value.value = va_arg( args, long );
					} else {
					value.value = va_arg( args, unsigned long );
					}
#endif
				} else if( longflag ){
					if( signed_val ){
						value.value = va_arg( args, long );
					} else {
						value.value = va_arg( args, unsigned long );
					}
				} else {
					if( signed_val ){
						value.value = va_arg( args, int );
					} else {
						value.value = va_arg( args, unsigned int );
					}
				}
				fmtnum( &value,plp_base,signed_val, ljust, len, zpad );
				break;
			case 's':
				strvalue = va_arg( args, char *);
				fmtstr(strvalue, ljust, len, precision);
				break;
			case 'c':
				ch = va_arg( args, int );
				{ char b[2];
					int vsb = visible_control;
					b[0] = ch;
					b[1] = 0;
					visible_control = 0;
					fmtstr(b, ljust, len, precision);
					visible_control = vsb;
				}
				break;
			case 'f': case 'g': case 'e':
				dval = va_arg( args, double );
				fmtdouble(ch, dval, ljust, len, zpad, precision);
				break;
			case 'm':
				fmtstr(plp_Errormsg(err), ljust, len, precision);
				break;
			case '%': dopr_outch( ch ); continue;
			default:
				dostr(  "???????" );
			}
			longflag = 0;
			break;
		default:
			dopr_outch( ch );
			break;
		}
	}
	*output = 0;
}

/*
 * Format '%[-]len[.precision]s'
 * -   = left justify (ljust)
 * len = minimum length
 * precision = numbers of chars in string to use
 */
static void
fmtstr(const char *value, int ljust, int len, int precision)
{
	int padlen, slen, i, c;	/* amount to pad */

	if( value == 0 ){
		value = "<NULL>";
	}
	if( precision > 0 ){
		slen = precision;
	} else {
		/* cheap slen so you do not have library call */
		for( slen = i = 0; (c=cval(value+i)); ++i ){
			if( visible_control && iscntrl( c ) && !isspace(c) ){
				++slen;
			}
			++slen;
		}
	}
	padlen = len - slen;
	if( padlen < 0 ) padlen = 0;
	if( ljust ) padlen = -padlen;
	while( padlen > 0 ) {
		dopr_outch( ' ' );
		--padlen;
	}
	/* output characters */
	for( i = 0; (c = cval(value+i)); ++i ){
		if( visible_control && iscntrl( c ) && !isspace( c ) ){
			dopr_outch('^');
			c = ('@' | (c & 0x1F));
		}
		dopr_outch(c);
	}
	while( padlen < 0 ) {
		dopr_outch( ' ' );
		++padlen;
	}
}

static void
fmtnum(  union value *value, int plp_base, int dosign, int ljust,
	int len, int zpad )
{
	int signvalue = 0;
#ifdef HAVE_LONG_LONG
	unsigned long long uvalue;
#else
	unsigned long uvalue;
#endif /* HAVE_LONG_LONG */
	char convert[64];
	int place = 0;
	int padlen = 0;	/* amount to pad */
	int caps = 0;

	/* fprintf(stderr,"value 0x%x, plp_base %d, dosign %d, ljust %d, len %d, zpad %d\n",
		value, plp_base, dosign, ljust, len, zpad );/ **/
	uvalue = value->value;
	if( dosign ){
		if( value->value < 0 ) {
			signvalue = '-';
			uvalue = -value->value;
		}
	}
	if( plp_base < 0 ){
		caps = 1;
		plp_base = -plp_base;
	}
	do{
		convert[place++] =
			(caps? "0123456789ABCDEF":"0123456789abcdef")
			 [uvalue % (unsigned)plp_base  ];
		uvalue = (uvalue / (unsigned)plp_base );
	}while(uvalue);
	convert[place] = 0;
	padlen = len - place;
	if( padlen < 0 ) padlen = 0;
	if( ljust ) padlen = -padlen;
	/* fprintf( stderr, "str '%s', place %d, sign %c, padlen %d\n",
		convert,place,signvalue,padlen); / **/
	if( zpad && padlen > 0 ){
		if( signvalue ){
			dopr_outch( signvalue );
			--padlen;
			signvalue = 0;
		}
		while( padlen > 0 ){
			dopr_outch( zpad );
			--padlen;
		}
	}
	while( padlen > 0 ) {
		dopr_outch( ' ' );
		--padlen;
	}
	if( signvalue ) dopr_outch( signvalue );
	while( place > 0 ) dopr_outch( convert[--place] );
	while( padlen < 0 ){
		dopr_outch( ' ' );
		++padlen;
	}
}

static void
plp_strcat(char *dest, const char *src )
{
	if( dest && src ){
		dest += strlen(dest);
		strcpy(dest,src);
	}
}

static void
fmtdouble( int fmt, double value, int ljust, int len, int zpad, int precision )
{
	char convert[128];
	char fmts[128];
	int l;

	if( len == 0 )
		len = 10;
	if( len > (int) (sizeof(convert) - 20) )
		len = sizeof(convert) - 20;
	if( precision > (int) sizeof(convert) - 20 )
		precision = sizeof(convert) - 20;
	if( precision > len )
		precision = len;
	strcpy( fmts, "%" );
	if( ljust )
		plp_strcat(fmts, "-" );
	if( zpad )
		plp_strcat(fmts, "0" );
	if( len )
		sprintf( fmts+strlen(fmts), "%d", len );
	if( precision > 0 )
		sprintf( fmts+strlen(fmts), ".%d", precision );
	l = strlen( fmts );
	fmts[l] = fmt;
	fmts[l+1] = 0;
	/* this is easier than trying to do the portable dtostr */
	sprintf( convert, fmts, value );
	dostr( convert );
}

static void dostr( const char *str )
{
	while(*str) dopr_outch(*str++);
}

static void dopr_outch( int c )
{
	if( end == 0 || output < end ){
		*output++ = c;
	}
}


/****************************************************************************
 * static char *plp_errormsg( int err )
 *  returns a printable form of the
 *  errormessage corresponding to the valie of err.
 *  This is the poor man's version of sperror(), not available on all systems
 *  Patrick Powell Tue Apr 11 08:05:05 PDT 1995
 ****************************************************************************/
/****************************************************************************/
#if !defined(HAVE_STRERROR)

# if defined(HAVE_SYS_NERR)
#  if !defined(HAVE_SYS_NERR_DEF)
     extern int sys_nerr;
#  endif
#  define num_errors    (sys_nerr)
# else
#  define num_errors    (-1)            /* always use "errno=%d" */
# endif

# if defined(HAVE_SYS_ERRLIST)
#  if !defined(HAVE_SYS_ERRLIST_DEF)
     extern const char *const sys_errlist[];
#  endif
# else
#  undef  num_errors
#  define num_errors   (-1)            /* always use "errno=%d" */
# endif

#endif

static char * plp_Errormsg ( int err )
{
    char *cp;

#if defined(HAVE_STRERROR)
	cp = (void *)strerror(err);
#else
# if defined(HAVE_SYS_ERRLIST)
    if (err >= 0 && err < num_errors) {
		cp = (void *)sys_errlist[err];
    } else
# endif
	{
		static char msgbuf[32];     /* holds "errno=%d". */
		/* SAFE use of sprintf */
		(void) sprintf(msgbuf, "errno=%d", err);
		cp = msgbuf;
    }
#endif
    return (cp);
}

#if defined(TEST)
#include <stdio.h>
int main( void )
{
	char buffer[128];
	char *t;
	char *test1 = "01234";
	errno = 1;
	plp_snprintf( buffer, sizeof(buffer), (t="errno '%m'")); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%s"), test1 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%12s"), test1 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%-12s"), test1 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%12.2s"), test1 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%-12.2s"), test1 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%g"), 1.25 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%g"), 1.2345 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%12g"), 1.25 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%12.2g"), 1.25 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%0*d"), 6, 1 ); printf( "%s = '%s'\n", t, buffer );
#if defined(HAVE_LONG_LONG)
	plp_snprintf( buffer, sizeof(buffer), (t = "%llx"), 1, 2, 3, 4 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%llx"), (long long)1, (long long)2 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%qx"), 1, 2, 3, 4 ); printf( "%s = '%s'\n", t, buffer );
	plp_snprintf( buffer, sizeof(buffer), (t = "%qx"), (quad_t)1, (quad_t)2 ); printf( "%s = '%s'\n", t, buffer );
#endif
	return(0);
}
#endif
#endif /* !HAVE_SNPRINTF && !HAVE_VSNPRINTF */
